CORE-4357 CPI Tool v2

kyriathar opened the pull request:
Adds CPI Tool for creating CPI v2 format 

(Code for this tool is copy and paste from v1 tool with needed minor modification, so it currently adds duplication, but we will be removing v1 tool so it should be ok).

- embeds the CPB instead of modifying it.
- adds CPI manifest main attributes as described [here](https://github.com/corda/platform-eng-design/blob/e494db1d6035c755095c20dcbebcce141b46adc2/core/corda-5/corda-5.0/packaging/cpi.md#manifest-attributes).
- adds tests asserting CPI v2 format.

Please note: it does not verify the CPB before packing it into the CPI. CPB verification for both CPI v1 and v2 tools will be added with CORE-5324.

josephzunigadaly commented on tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/TestUtils.kt:
```diff
@@ -139,11 +140,33 @@ internal object TestUtils {
     fun jarEntriesExistInCpx(cpxPath: Path, expectedEntries: List<String>): Boolean {
-        JarInputStream(Files.newInputStream(cpxPath)).use {
-            val actualEntries = generateSequence { it.nextJarEntry?.name }.toList()
+        JarInputStream(Files.newInputStream(cpxPath)).use { jarStream ->
+            val actualEntries = generateSequence {
+                jarStream.nextJarEntry?.name
+            }.toMutableList()
+
+            jarStream.manifest?.let {
+                actualEntries += "META-INF/MANIFEST.MF"
+            }
```
If we replaced `JarInputStream` with `ZipInputStream` in the original code, would that have the same result?

kyriathar replied:
That should be fine for jar entries since we only use their names in this case so `ZipEntry` would be equally good. However, we also grab `Manifest` so I believe no.

josephzunigadaly replied:
Manifest gets special treatment in `JarInputStream` but not `ZipInputStream`. The manifest would appear as another `ZipEntry`. There would be no need for the extra code to check for manifest when using `ZipInputStream`.

kyriathar replied:
Done

josephzunigadaly commented on tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/TestUtils.kt:
```diff
@@ -139,11 +140,33 @@ internal object TestUtils {
         }
     }
+
+    fun containsAllManifestAttributes(
+        cpxFile: Path,
+        expectedManifestAttributes: Map<Attributes.Name, String>
+    ): Boolean =
+        JarInputStream(Files.newInputStream(cpxFile)).use { stream ->
+            val actualManifestAttributes = stream.manifest.mainAttributes
+            expectedManifestAttributes.all {
+                val attributeName = it.key
+                val attributeValue = it.value
+                actualManifestAttributes[attributeName] != null &&
+                        actualManifestAttributes[attributeName] == attributeValue
+
+            }
+        }
```
How do we know which attribute was missing when this fails?

A quick fix would be to change this from a true/false return to throw an exception with a good message on failure. An assertEquals within the method would do the job and give a good error message.

There are other solutions possible too. The aim would be to fully understand the cause of a failure by looking at the test result. No debugging.

kyriathar replied:
I wanted to try and not have asserts in these methods in here, just them to return true/ false. Good point, will address this.

kyriathar-r3 committed:
CORE-4357 CPI Tool v2 (#1774)

Adds CPI Tool for creating CPI v2 format.

- embeds the CPB instead of modifying it.
- adds CPI manifest main attributes.

mnesbit reviewed (approved):
OK
